### PR TITLE
Add ability to render sidecar templates for captions and content after a form definition

### PR DESCRIPTION
**Pull request by camertron**
I'm working on a form in dotcom right now where the caption text for certain fields is complicated enough that it's awkward to write in Ruby code. It would instead be really nice if captions could be rendered from templates. In addition, certain forms contain invisible content (eg. popovers) that needs to be rendered on the page somewhere. To address both these use-cases, I propose introducing sidecar templates.

### Caption templates

Consider the following form:

```ruby
class MyForm < ApplicationForm
  form do |my_form|
    my_form.text_field(
      name: :first_name,
      label: "First name"
    )
  end
end
```

If no caption is provided, PRF will look for a template called first_name_caption.html.erb in app/forms/my_form/ and render it where the caption would normally go.

### After content

Consider the example above. If a template called after_content.html.erb exists in app/forms/my_form/ it will be rendered after all the form inputs.

### CODE REVIEW

**Comment by langermank** [comment]
Is one of the use cases rendering additional labeling/caption context in a tooltip? I wonder if you might want to consult with the accessibility team on that case.

**Comment by camertron** [comment]
@langermank I'm not specifically targeting tooltips at the moment, but a question did come up in our discussions wrt accessibility of popovers and if it makes sense to support explainer text per-input. The solution in this PR is designed to let you render _any_ content after the form, which can include popovers (or anything else). We'll have to go through and specifically audit any forms that currently do this.

File: lib/primer/rails_forms/acts_as_component.rb [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+      module InstanceMethods
+        delegate :render, :content_tag, :output_buffer, :capture, to: :@view_context
+
+        def render_in(view_context, &block)
+          @view_context = view_context
+          perform_render(&block)
+        end
+
+        def perform_render(&_block)
+          # override in host class
```
Comment (joelhawksley) [inline lib/primer/rails_forms/acts_as_component.rb:15]: Would we want to raise a NotImplementedError here then?

**Review by joelhawksley — COMMENTED** [review]
Looking good! Let's 🍐   review this today ❤️

File: lib/primer/rails_forms/acts_as_component.rb [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+      def self.extended(base)
+        base.include(InstanceMethods)
+      end
+
+      TemplateGlob = Struct.new(:glob_pattern, :method_name, :on_compile_callback)
+      TemplateParams = Struct.new(:source, :identifier, :type, :format)
+
+      attr_accessor :template_root_path
+
+      def renders_templates(glob_pattern, &block)
```
Comment (camertron) [inline lib/primer/rails_forms/acts_as_component.rb:28]: Can one `renders_*` method call the other one?

File: lib/primer/rails_forms/base.html.erb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<% each_input do |input| %>
+  <%= render(input.to_component(builder: @builder, form: self)) %>
+<% end %>
+<% if after_content? %>
+  <%= render_after_content %>
```
Comment (camertron) [inline lib/primer/rails_forms/base.html.erb:5]: Consider making these auto-generated methods private or prefixing with `__` perhaps?

File: lib/primer/rails_forms/caption.html.erb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+<% if @context.caption? %>
```
Comment (camertron) [inline lib/primer/rails_forms/caption.html.erb:1]: Let's raise an error if there are both a caption passed in the context _and_ a caption template.

File: lib/primer/rails_forms/form_control.html.erb [hunk 5]
```diff
@@ -15,6 +15,4 @@
     <span><%= @context.validation_messages.first %></span>
   </div>
 <% end %>
-<% if @context.caption? %>
-  <p class="note" id="<%= @context.caption_id %>"><%= @context.caption %></p>
-<% end %>
+<%= render(Caption.new(input: @input, form: @form, context: @context)) %>
```
Comment (camertron) [inline lib/primer/rails_forms/form_control.html.erb:18]: Don't forget to do this for check boxes and radio buttons too :)

**Pull request merged**